### PR TITLE
Implement Ransack Gem

### PR DESCRIPTION
**Pull request by Eman-alvi29**

### CODE REVIEW

File: Ecommerce/app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,7 +1,14 @@
 class ApplicationController < ActionController::Base
+  before_action :set_ransack_search
+
   def after_sign_out_path_for(resource_or_scope)
     new_user_session_path # Redirect to the login page
   end
 
-   allow_browser versions: :modern
+  allow_browser versions: :modern
+  private
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/application_controller.rb]: add an extra line before this line.

File: Ecommerce/app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,7 +1,14 @@
...
+
+  def set_ransack_search
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/application_controller.rb]: give a tab space to all in private method

File: Ecommerce/app/controllers/application_controller.rb [hunk 3]
```diff
@@ -1,7 +1,14 @@
...
+  before_action :set_ransack_search
+
   def after_sign_out_path_for(resource_or_scope)
     new_user_session_path # Redirect to the login page
   end
 
-   allow_browser versions: :modern
+  allow_browser versions: :modern
+  private
+
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/application_controller.rb]: Remove this extra line

File: Ecommerce/app/controllers/categories_controller.rb [hunk 4]
```diff
@@ -1,7 +1,7 @@
 class CategoriesController < ApplicationController
   def index
-    @category = params[:category]
-    @items = @category.present? ? Item.where(category: @category) : Item.all
+    @q = Item.ransack(params[:q])
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/categories_controller.rb]: in categories controller why we are applying ransack on item?

Comment (Eman-alvi29) [reply Ecommerce/app/controllers/categories_controller.rb]: When I created the Items controller, it wasn't being used, so you told me to delete it. Then, I displayed the items in the Categories controller, that is why I used Ransack there.

Comment (asadsarfraz2347) [reply Ecommerce/app/controllers/categories_controller.rb]: But now is the time to use the items controller,  you can not write the items controller index method in categories controller index method.
Whenever I told you to remove or add something it has some purpose. But the purpose does not change the rules for future developement,

File: Ecommerce/app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -13,71 +13,77 @@
...
+        </button>
 
-        <li class="nav-item">
-          <a class="nav-link" href="#">About Us</a>
-        </li>
+        <div class="collapse navbar-collapse" id="navbarSupportedContent">
+          <ul class="navbar-nav me-auto mb-2 mb-lg-0">
+            <li class="nav-item">
+              <a class="nav-link active" aria-current="page" href="#">Home</a>
+            </li>
```
Comment (asadsarfraz2347) [inline Ecommerce/app/views/layouts/application.html.erb]: add extra line after this

File: Ecommerce/app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -13,71 +13,77 @@
...
+            <li class="nav-item">
+              <a class="nav-link" href="#">Products</a>
+            </li>
```
Comment (asadsarfraz2347) [inline Ecommerce/app/views/layouts/application.html.erb]: add extra line after this

File: Ecommerce/app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -13,71 +13,77 @@
...
+            <li class="nav-item">
+              <a class="nav-link" href="#">Categories</a>
+            </li>
```
Comment (asadsarfraz2347) [inline Ecommerce/app/views/layouts/application.html.erb]: add extra line after this

File: Ecommerce/app/views/layouts/application.html.erb [hunk 8]
```diff
@@ -13,71 +13,77 @@
...
+            <li class="nav-item">
+              <a class="nav-link" href="#">About Us</a>
+            </li>
```
Comment (asadsarfraz2347) [inline Ecommerce/app/views/layouts/application.html.erb]: add extra line after this

File: Ecommerce/app/views/layouts/application.html.erb [hunk 9]
```diff
@@ -13,71 +13,77 @@
...
 
-      </ul>
-      <form class="d-flex me-3">
-        <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
-        <button class="btn btn-outline-light" type="submit">Search</button> <!-- Light button color -->
-      </form>
+          <div class="dropdown me-3">
+            <a class="nav-link dropdown-toggle" href="#" id="profileDropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
+              <img src="path/to/user-profile-pic.jpg" alt="User" class="rounded-circle" style="height: 30px; width: 30px;">
+            </a>
```
Comment (asadsarfraz2347) [inline Ecommerce/app/views/layouts/application.html.erb]: add extra line after this

File: Ecommerce/app/controllers/application_controller.rb [hunk 10]
```diff
@@ -1,7 +1,13 @@
 class ApplicationController < ActionController::Base
+  before_action :set_ransack_search
+
   def after_sign_out_path_for(resource_or_scope)
     new_user_session_path # Redirect to the login page
   end
 
-   allow_browser versions: :modern
+  private
+
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/application_controller.rb]: Remove this extra line

File: Ecommerce/app/controllers/items_controller.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+    @items = @q.result(distinct: true)
+    @categories = Category.all
+  end
+
+  def show
+    @item = Item.find(params[:id])
+  end
+
+  private
+  def set_ransack_search
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/items_controller.rb]: give tab space to this method

File: Ecommerce/app/controllers/welcome_controller.rb [hunk 12]
```diff
@@ -2,6 +2,6 @@ class WelcomeController < ApplicationController
   before_action :authenticate_user!
 
   def index
-      @items = Item.all
-  end
+   @q = Item.ransack(params[:q])
```
Comment (asadsarfraz2347) [inline Ecommerce/app/controllers/welcome_controller.rb]: if we want to show items, why we are not going into items controller? why we are hitting welcome controller?

**Pull request merged**